File: src/main/java/com/victorc/gymnotebook/models/Exercise.java
package com.victorc.gymnotebook.models;

import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "exercises")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Exercise {
	
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@NotNull
	private String name;

	@OneToOne
	private ImageData image;

	private String description;

	@NotNull
	@Enumerated(EnumType.STRING)
	private EExerciseType type;

	@OneToMany(mappedBy = "exercise", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
	private List<WorkoutSet> workoutSets;

	@NotNull
	@Enumerated(EnumType.STRING)
	private EMuscleGroup primaryMuscleGroup;

	@Enumerated(EnumType.STRING)
	private EMuscleGroup secondaryMuscleGroup;

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "user_id", nullable = false)
	private User user;
}

File: apps/api/src/exercises/domain/exercise.ts
import type { EExerciseType, EMuscleGroup } from '@gymnotebook/contracts';

export type { EExerciseType, EMuscleGroup };

export interface Exercise {
  id: number;
  name: string;
  description: string | null;
  imageId: number | null;
  type: EExerciseType;
  primaryMuscleGroup: EMuscleGroup;
  secondaryMuscleGroup: EMuscleGroup | null;
  userId: number;
}

export interface ExerciseDraft {
  name: string;
  description: string | null;
  imageId: number | null;
  type: EExerciseType;
  primaryMuscleGroup: EMuscleGroup;
  secondaryMuscleGroup: EMuscleGroup | null;
  userId: number;
}

export type ExerciseUpdate = Omit<ExerciseDraft, 'userId'>;

export function normalizeExerciseInput<T extends Omit<ExerciseDraft, 'userId'>>(input: T): T {
  return {
    ...input,
    name: input.name.trim(),
    description: normalizeOptionalText(input.description),
  };
}

function normalizeOptionalText(value: string | null): string | null {
  if (value === null) {
    return null;
  }
  const trimmed = value.trim();
  return trimmed.length === 0 ? null : trimmed;
}
